code/Controllers/HomeController.cs

```diff
@@ -5,79 +5,110 @@
-		}
-	}
+    public class HomeController : Controller
+    {
+        private static readonly Regex productRegex = new Regex(@"product\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+        private static readonly Regex tagRegex = new Regex(@"tags\/(.*?)\/*?$", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+        private static readonly Regex minPriceRegex = new Regex(@"minprice\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+        private static readonly Regex maxPriceRegex = new Regex(@"maxprice\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+        private static readonly Regex aboutRegex = new Regex(@"aboutus", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+
+        private readonly BagsApi bagsApi;
+
+        public HomeController(BagsApi bagsApi)
+        {
+            this.bagsApi = bagsApi;
+        }
+
+        [Route("")]
+        [Route("app/{*path}")]
+        public async Task<IActionResult> App(String path = null)
+        {
+            var model = new MetaViewModel();
+
+            path = path ?? "";
+            var productId = productRegex.Match(path).Groups[1].Value?.TryParseUInt64();
+            var tagIds = tagRegex.Match(path).Groups[1].Value;
+            var minPrice = minPriceRegex.Match(path).Groups[1].Value.TryParseUInt64();
+            var maxPrice = maxPriceRegex.Match(path).Groups[1].Value.TryParseUInt64();
+            var aboutUs = aboutRegex.Match(path).Success;
+
+            if (productId != null)
+            {
+                var product = await bagsApi.GetProduct(productId ?? 0);
+
+                model.Url = $"https://bagcupid.com/app/{path.TrimStart('/')}";
+
+                if (product?.Images != null && product.Images.Count() > 0)
+                    model.Images = product.Images
+                        .OrderBy(image => image.Priority)
+                        .SelectAndSwallowReferenceType(image => new Uri(image.Large))
+                        .Where(image => image != null);
+
+                if (product?.Tags != null && product.Tags.Count() > 0)
+                {
+                    var brand = product.Tags.FirstOrDefault(x => x.IsBrand)?.TagName ?? "";
+                    var styles = String.Join("/", product.Tags.Where(x => x.IsStyle).Select(x => x.TagName).ToArray());
+
+                    model.Title = $"{brand} {product.Name} - {styles}";
+                }
+            }
+            else if (!String.IsNullOrEmpty(tagIds))
+            {
+                model.Url = $"https://bagcupid.com/app/{path.TrimStart('/')}";
+
+                var tagList = tagIds.Split('_').ToList();
+                var productsTask = bagsApi.GetProductsByTags(tagList);
+                var taskTagModels = bagsApi.GetTags();
+
+                var products = await productsTask;
+                if (products?.Count > 0)
+                    model.Images = (await productsTask ?? Enumerable.Empty<BagsApi.Product>())
```

Comment:
This should be `model.Images = (products ?? Enumerable.Empty<BagsApi.Product>())` since you already have awaited the task above.

Also, you can avoid dealing with nulls by moving the null coalesce up a bit:
```
var products = (await productsTask) ?? Enumerable.Empty<BagsApi.Product>();
if (products.Count > 0)
   ...
```

Lastly, because the code inside the if block is lengthy, it should include curly braces to make it clear where the beginning/end is.